ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java

```diff
@@ -80,6 +80,7 @@ public boolean accept(File dir, String name) {
   final ConcurrentSkipListMap<Offset, IndexSegment> indexes = new ConcurrentSkipListMap<>();
   final Journal journal;
   final HardDeleter hardDeleter;
+  volatile Thread hardDeleteThread = null;
```

Comment:
`volatile` is required only in certain special situations involving concurrency.

I would say it can still be made `final`. All you need to do is set it to `null` in the else of `isHardDeleteEnabled`.